// --------------------------------------------------------
// WARNING: XVIZ Protobuf files are unstable, do not use
// --------------------------------------------------------

syntax = "proto3";

package xviz.v2;

option csharp_namespace = "xviz.V2";
option go_package = "v2pb";
option java_multiple_files = true;
option java_outer_classname = "PrimitivesProto";
option java_package = "com.xviz.v2";
option objc_class_prefix = "XVIZ";
option php_namespace = "Xviz\\V2";

import "xviz/v2/options.proto";
import "xviz/v2/style.proto";

message PrimitiveBase {
  //    option (xviz_json_schema) = "primitives/base";

  string object_id = 1;
  repeated string classes = 2;
  StyleObjectValue style = 3;
}

message Circle {
  option (xviz_json_schema) = "primitives/circle";
  PrimitiveBase base = 1;
  repeated double center = 2;
  double radius = 3;
}

message Image {
  option (xviz_json_schema) = "primitives/image";
  PrimitiveBase base = 1;
  repeated double position = 2;
  bytes data = 3;
  uint32 width_px = 4;
  uint32 height_px = 5;
}

message Point {
  option (xviz_json_schema) = "primitives/point";
  PrimitiveBase base = 1;
  // Flattened X, Y, Z list
  repeated float points = 2;
  // Flattened list of (R, G, B) or (R, G, B, A)
  repeated uint32 colors = 3;

  // 3, 32 bit floats per point
  bytes points_buffer = 4;
  // 4, 8 bit values per value
  bytes colors_buffer = 5;
}

message Polygon {
  option (xviz_json_schema) = "primitives/polygon";
  PrimitiveBase base = 1;
  // Flattened X, Y, Z list
  repeated float vertices = 2;
}

message Polyline {
  option (xviz_json_schema) = "primitives/polyline";
  PrimitiveBase base = 1;
  // Flattened X, Y, Z list
  repeated float vertices = 2;
  // Flattened (R, G, B) or (R, G, B, A) list
  repeated uint32 colors = 3;
}

message PolylineGroup {
  // Totally un-optimized style and ID information
  // TODO(jlisee): have a columnized version of this
  repeated PrimitiveBase bases = 1;

  // 1, 16 bits unsigned int per polyline
  bytes vertex_counts_buffer = 2;
  // 3, 32 bit floats per point (X, Y, Z)
  repeated float vertices = 3;
}

message Stadium {
  option (xviz_json_schema) = "primitives/stadium";
  PrimitiveBase base = 1;
  repeated double start = 2;
  repeated double end = 3;
  double radius = 4;
}

message Text {
  option (xviz_json_schema) = "primitives/text";
  PrimitiveBase base = 1;
  repeated double position = 2;
  string text = 3;
}
